lib/timsort_lib_purecap.c

```diff
@@ -0,0 +1,237 @@
+#include "timsort_lib_purecap.h"
+#include "../include/common.h"
+#include <assert.h>
+#include <cheri/cheric.h>
+#include <cheriintrin.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+const int RUN_LENGTH = 32;
+
+/**
+ * Checks if an array of integers `arr` is sorted in ascending order. Uses capability instructions
+ * to determine the array's length.
+ * @param arr Array to sort
+ * Capability implicit paramters:
+ * - uses length of memory allocation chunk as upper bound (unit: bytes)
+ */
+bool isSorted(int *arr)
+{
+	assert(cheri_is_valid(arr));
+	size_t length = cheri_getlen(arr) / sizeof(int);
+
+	// short-circuit: empty and singleton arrays are always sorted.
+	// still makes sense in capability land as lengths <= 1024 are exact
+	if (length <= 1)
+	{
+		return true;
+	}
+
+	for (size_t ix = 1; ix < length; ix++)
```

Comment:
Does `cheri_getlen` guarantee to always return precisely the correct length of the array? The comment above:

> still makes sense in capability land as lengths <= 1024 are exact

suggests "no", in which case I think this loop might end up checking incorrect amounts of data (presumably too much rather than too little?)?

Reply:
That is correct on both counts: it does not return precisely the length of the array and it may end up checking too much data. I referenced this issue in the first PR description comment ( at the top ) where i mentioned that the length of the array is hardcoded to a "known to be exact" value of 8192. 

The "correct way" that i can imagine to handle the general case implies a dynamic switch between the two approaches to parameter passing.  Ibe would 
 - use `cheri_bounds_set_exact` and check the return value ( for the use in timsort_lib_purecap.c:231 for example ) to predicate a call to either `void merge(int *arr)`  (from timsort_lib_purecap.c) or `void merge(int arr[], size_t lowerBound, size_t midPoint, size_t upperBound)`   (from timsort_lib.c)  when the bound can't be exactly represented 
 - check the requested allocation size versus what bounds are reported by `cheri_getlen` and make the same dynamic decision.

The purpose of this example in my mind is to surface such issues so I don't believe the additional engineering effort is worth much at present but am happy to hear different opinions or alternative implementation plans to remove this hardcoded assumption.

Reply:
OK, so what I've learnt from this is that we can't use normal CHERI bounds in place of precise knowledge of how big an array is :) Does `cheri_bounds_set_exact` guarantee to precisely set the bounds and can we read back that precise value?

Reply:
Indeed that's a good conclusion to draw. It may still be a useful trick when the array size is statically known to be small, or more generally, representable. This was to be expected given that the `bounds` field is by necessity lot smaller than the `address` field. The question i have left remaining here is whether this  ( having loads of small arrays ) is a useful pattern that can be seen "in the wild". 

>  Does cheri_bounds_set_exact guarantee to precisely set the bounds  

It looks like the only thing `cheri_bounds_set_exact` buys us is a hardware exception if it does not succeed in setting the exact value: 

> "This  is  similar  to cheri_bounds_set() but  will raise a hardware exception if the resulting capability is not precisely representable in-stead of rounding the bounds. "

So the only remaining solution is to use `cheri_bounds_set` and check with `cheri_getlen`.

Reply:
> So the only remaining solution is to use cheri_bounds_set and check with cheri_getlen.

I guess these still can't *exactly* represent all possible lengths?

Reply:
Correct. There's no way that I know of to represent an exact length in the general case.

Reply:
Got it.

What we could do, maybe, is have a version of timsort called something like `timsort_cheri_bounds` which has a big warning at the top saying something like "WARNING: THIS IS ONLY CORRECT FOR BOUNDS WHICH CHERI CAN PRECISELY REPRESENT." And then have another, guaranteed correct, version which has slightly worse bounds checking (in some senses) but is actually correct. Sound like a plan?

Reply:
Sounds like a plan indeed! I think extra safety checking makes for a more convincing example to boot. 

I will: 
- insert a warning and rename the _purecap version ( perhaps to _purecap_unsafe) 
- add a "hybrid" version that performs the required dynamic dispatch

Reply:
Cool!

Reply:
This is now implemented. Beware, running `make run-test-timsort_purecap` will spin the CPU for a while.